packages/react-native/ReactCommon/jsc/JSCRuntime.cpp

```diff
@@ -290,11 +290,16 @@ class JSCRuntime : public jsi::Runtime {
     }                          \
   } while (0)
 
-#if defined(__IPHONE_OS_VERSION_MIN_REQUIRED)
+#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED)
 #if __IPHONE_OS_VERSION_MAX_ALLOWED >= 160400
 #define _JSC_HAS_INSPECTABLE
 #endif
 #endif
+#if defined(__MAC_OS_X_VERSION_MAX_ALLOWED)
+#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 130300
```

Comment:
```suggestion
#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 13300
```

If MacOS follows the same semantics of iOS

Reply:
In `AvailabilityVersions.h`:

```
#define __IPHONE_16_4                                   160400
#define __IPHONE_16_5                                   160500
#define __IPHONE_16_6                                   160600
#define __IPHONE_16_7                                   160700
#define __IPHONE_17_0                                   170000
#define __IPHONE_17_1                                   170100
#define __IPHONE_17_2                                   170200
#define __IPHONE_17_3                                   170300
#define __IPHONE_17_4                                   170400
```

Reply:
uhm... so, why the check `#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 160400` failed on devices < 16.4? 😱

Reply:
I'll do some testing with an iOS 15 simulator and try to update. Luckily Xcode has a way to see post macro processed files (though I forget the command at this time)

Reply:
I did some testing, and as far as I can tell. the existing code is right. 
With local testing, I added two printf statements to the offending block and tested on an iOS 15.5 and iOS 17.2 Simulator + RNTester with `USE_HERMES=0 pod install`:
```cpp
//#if (__OSX_AVAILABLE_STARTING(MAC_13_3, IPHONE_16_4))
  printf("Inside JSC_HAS_INSPECTABLE hit");
  if (__builtin_available(macOS 13.3, iOS 16.4, tvOS 16.4, *)) {
    printf("Inside __builtin_available hit");
    JSGlobalContextSetInspectable(ctx_, true);
  }
//#endif
#endif
```

With the outer `__OSX_AVAILABLE_STARTING` macro commented out, iOS 17.2 will hit both printf statements, while iOS 15.5 will only hit the outer statement (expected). However, with the extra `__OSX_AVAILABLE_STARTING` commented in, we hit neither printf statement on both iOS 15.5 and iOS 17.2! I can confirm this breaks JSC direct debugging on iOS 16.4+ as well 😨 .


Looking even deeper at `Availability.h`, I suspect the original crash repoprted the apple forum had their app build setting misconfigured. I see this in comments around the `__OSX_AVAILABLE_STARTING ` macro:

> For these macros to function properly, a program must specify the OS version range 
    it is targeting.  The min OS version is specified as an option to the compiler:
    -mmacosx-version-min=10.x when building for Mac OS X, and -miphoneos-version-min=y.z
    when building for the iPhone.  The upper bound for the OS version is rarely needed,
    but it can be set on the command line via: -D__MAC_OS_X_VERSION_MAX_ALLOWED=10x0 for
    Mac OS X and __IPHONE_OS_VERSION_MAX_ALLOWED = y0z00 for iOS.

Reply:
But we were having an app crashing internally because of that.. 🤔 and it's hard to imagine that the internal app was misconfigured in the same way...
So, how's that possible?

Reply:
Really? There was an internal app using JSC? Now I don't know 😵‍💫

Reply:
yes, unfortunately some teams are still on JSC... 🤷

Reply:
(Unless of course my testing was wrong and the patch for the new macro is actually fine.. maybe I'll try outside of RNTester

Reply:
Update:

- The crash only happens on device, hence we didn't catch it in simulator (Thankfully I kept my old iPad)
- System frameworks are dynamically linked, so the symbol lookup happens at runtime. That makes sense why it's a crash and not a failure to compile.
- It can be fixed by linking `JavascriptCore` as a weak_framework instead of a framework. Doing so means it will explicitly link all the sumbols in the framework weakly, and seems to be the intended way to deal with new APIs. We don't use weak frameworks internally much, so it's a bit of a blind spot to me. https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WeakLinking.html

I updated this PR so that we weakly link JavascriptCore. I also totally changed the title and description because the scope is now _very_ different :D